features/steps/givens/attributes.py

```diff
@@ -74,27 +72,24 @@ def step_impl(context, inst, attribute, condition, prefix):
     elif condition == 'does not start':
         if hasattr(inst, attribute) and not str(getattr(inst, attribute, '')).startswith(prefix):
             yield ValidationOutcome(instance_id=inst, severity=OutcomeSeverity.PASSED)
-            
-
-@gherkin_ifc.step('Its final segment')
-@gherkin_ifc.step('Its final segment at depth 1')
-def step_impl(context, inst):
-    """
-    Implement ALS015
-    This is a separate function from ALB015 because ALS015 needs to return an entity_instance.
-    """
-    yield ValidationOutcome(instance_id = inst[-1], severity=OutcomeSeverity.PASSED)
-
-
-@gherkin_ifc.step('Its final IfcAlignmentSegment')
-def step_impl(context, inst):
-    """
-    Implement ALB015
-    This is a separate function from ALS015 because ALB015 needs to yield a ValidationOutcome.
-    """
-    yield ValidationOutcome(instance_id=context.instances[-1], severity=OutcomeSeverity.PASSED)
 
 
+@gherkin_ifc.step('Its {ff:first_or_final} element')
+@gherkin_ifc.step('Its {ff:first_or_final} element at depth 1')
+@gherkin_ifc.step('Its {ff:first_or_final} element filtering by {filter}')
+@gherkin_ifc.step('Its {ff:first_or_final} element at depth 1 filtering by {filter}')
+def step_impl(context, inst, ff=0, filter=None):
+    def get_filter_key(attr):
+        if attr == 'id':
+            return lambda entity: entity.id() # in case of a method
+        else:
+            return lambda entity: getattr(entity, attr, None) # in case of an attribute
+
+    if ff:
+        yield ValidationOutcome(instance_id =  max(inst, key=get_filter_key(filter)) if filter else inst[-1], severity=OutcomeSeverity.PASSED)
```

Comment:
I don't understand the intention behind this `... filtering by id`.

Based on that you're taking the instance with the highest or lower (`min`, `max` respectively using a key transformation of inst -> inst.id() [which doesn't exist on non-instances]). But the id is only a local naming mechanism in the exchange format, it really should not have any significance outside of the exchange format.

Reply:
```
    Given A model with Schema "IFC4.3"                                                               # features/steps/givens/attributes.py:45 0.002s
    Given An IfcAlignmentCant                                                                        # features/steps/givens/entities.py:12
[#2388=IfcAlignmentCant('0kwyzC0yb8ehcYFOVppIP9',#1,'C1',$,$,#2389,$,1.435)]

    Given A relationship IfcRelNests from IfcAlignmentCant to IfcAlignmentSegment and following that # features/steps/givens/relationships.py:13
[[#2435=IfcAlignmentSegment('33RHONOYr9mBgFO3_aTR9K',#1,'c8',$,$,#2436,#2680,#2440), #2441=IfcAlignmentSegment('0v2fCSQ8H1qejZRDkDUoz1',#1,'c9',$,$,#2442,#2691,#2446), #2393=IfcAlignmentSegment('0o7VCJvavDdA9xYwz5oBRK',#1,'c1',$,$,#2394,#2603,#2398), #2399=IfcAlignmentSegment('2pwzDsynP3ggq3OBYyg7wr',#1,'c2',$,$,#2400,#2614,#2404), #2405=IfcAlignmentSegment('2tLO9I_A1DbQwnSyol1_P8',#1,'c3',$,$,#2406,#2625,#2410), #2411=IfcAlignmentSegment('1CWfkLVFXCo8JV8zZdrLfo',#1,'c4',$,$,#2412,#2636,#2416), #2417=IfcAlignmentSegment('08LKrvnvj079Xxjd3KbIy9',#1,'c5',$,$,#2418,#2647,#2422), #2423=IfcAlignmentSegment('1_39nu3q10EgiUF2uuPA6u',#1,'c6',$,$,#2424,#2658,#2428), #6001=IfcAlignmentSegment('1v2fCSQ8H1qejZRDkDUoz2',#1,'cA',$,$,#2442,#6101,#6011), #2429=IfcAlignmentSegment('2e6W3TPHX9jAKv_l1faiaw',#1,'c7',$,$,#2430,#2669,#2434)]]

    Given Its final element at depth 1                                                               # features/steps/givens/attributes.py:77
[#2429=IfcAlignmentSegment('2e6W3TPHX9jAKv_l1faiaw',#1,'c7',$,$,#2430,#2669,#2434)]

    Given Its attribute DesignParameters                                                             # features/steps/givens/attributes.py:62
[#2434=IfcAlignmentCantSegment($,$,740.169262678128,109.431749924283,3.E-2,3.E-2,-3.E-2,-3.E-2,.CONSTANTCANT.)] 

    Then The HorizontalLength of the final IfcAlignmentCantSegment must be 0                         # features/steps/thens/attributes.py:94 0.001s
      Assertion Failed: Errors occured:
      ["Feature=ALB015 Outcome=E00020 Severity=ERROR Expected={'value': 0.0} Observed={'value': 109.431749924283}"]

```

^ `#2429` is indeed the last segment of the list

What was the filtering by id intended to do?

Reply:
Filtering by id was being used to select the last segment of the list because in our test files the STEP id is always increasing as you traverse along the alignments.  As you noted above the id is not semantically significant and not guaranteed to always be in increasing order along the alignment.

`Its final element at depth 1` also aligns nicely with ALS015 which is the companion to ALB015.